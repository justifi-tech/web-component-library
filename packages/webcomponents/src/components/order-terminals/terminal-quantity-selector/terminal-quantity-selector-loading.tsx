import { h } from '@stencil/core';
import { Skeleton } from '../../../ui-components';

export const TerminalSelectorLoading = () => {
  return (
    <div class="terminal-selector-skeleton mt-4 rounded shadow-sm">
      <div class="d-flex align-items-center gap-3">
        <Skeleton width="125px" height="125px" styles={{ borderRadius: '8px' }} />

        <div class="flex-grow-1">
          <div class="d-flex row flex-row">
            <div class="col-6 flex-column">
              <Skeleton />
              <Skeleton width="60%" />
            </div>

            <div class="d-flex justify-content-end align-items-center gap-2 col-6">
              <Skeleton width="32px" height="32px" styles={{ borderRadius: '50%' }} />
              <Skeleton width="80px" height="32px" />
              <Skeleton width="32px" height="32px" styles={{ borderRadius: '50%' }} />
            </div>
          </div>
          <div class="mt-4">
            <div class="d-flex flex-column gap-2">
              <Skeleton />
              <Skeleton />
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
