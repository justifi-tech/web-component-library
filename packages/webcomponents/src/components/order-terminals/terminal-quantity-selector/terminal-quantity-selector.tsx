import { Component, Event, EventEmitter, h, Prop, State } from "@stencil/core";
import { buttonLink, image, text } from "../../../styles/parts";
import { TerminalModelName } from "../../../api";

@Component({
  tag: 'terminal-quantity-selector',
  shadow: false,
})
export class TerminalQuantitySelector {
  @Prop() modelName: TerminalModelName;
  @Prop() imageUrl: string;
  @Prop() description: string;
  @Prop() helpUrl: string;
  @Prop() limit: number = Infinity;

  @State() selectedQuantity: number = 0;

  @Event() selectedQuantityChange: EventEmitter<{ modelName: TerminalModelName, quantity: number }>;

  private removeUnit = () => {
    if (this.selectedQuantity > 0) {
      this.selectedQuantity--;
      this.selectedQuantityChange.emit({ modelName: this.modelName, quantity: this.selectedQuantity });
    }
  }

  private addUnit = () => {
    if (this.limit > 0) {
      this.selectedQuantity++;
      this.selectedQuantityChange.emit({ modelName: this.modelName, quantity: this.selectedQuantity });
    }
  }

  render() {
    return (
      <div class="rounded shadow-sm card p-2" part={text}>
        <div class="d-flex gap-3">
          <div class="d-flex align-items-center">
            <img src={this.imageUrl} alt={this.modelName} height={75} width={75} part={image} />
          </div>

          <div class="d-flex flex-column flex-grow-1 gap-3 p-4 justify-content-center">
            <div class="row">
              <div class="col-6 d-flex gap-2">
                {this.modelName}
                <form-control-tooltip helpText={this.description} />
              </div>
              <div class="col-6 d-flex justify-content-end align-items-center">
                <button class="p-0 minus btn" onClick={this.removeUnit} part={buttonLink}>-</button>
                <div class="w-25 text-center">{this.selectedQuantity}</div>
                <button class="p-0 plus btn" onClick={this.addUnit} part={buttonLink}>+</button>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
