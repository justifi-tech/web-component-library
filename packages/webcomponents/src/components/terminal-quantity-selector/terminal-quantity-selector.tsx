import { Component, Event, EventEmitter, h, Prop, State } from "@stencil/core";
import { buttonLink, image, link, text } from "../../styles/parts";
import { TerminalModelName } from "../../api";

@Component({
  tag: 'terminal-quantity-selector',
  shadow: false,
})
export class TerminalQuantitySelector {
  @Prop() modelName: TerminalModelName;
  @Prop() imageUrl: string;
  @Prop() description: string;
  @Prop() helpUrl: string;
  @Prop() limit: number = Infinity;

  @State() selectedQuantity: number = 0;

  @Event() selectedQuantityChange: EventEmitter<{ modelName: TerminalModelName, quantity: number }>;

  private removeUnit = () => {
    if (this.selectedQuantity > 0) {
      this.selectedQuantity--;
      this.selectedQuantityChange.emit({ modelName: this.modelName, quantity: this.selectedQuantity });
    }
  }

  private addUnit = () => {
    if (this.limit > 0) {
      this.selectedQuantity++;
      this.selectedQuantityChange.emit({ modelName: this.modelName, quantity: this.selectedQuantity });
    }
  }

  render() {
    return (
      <div class="mt-4 rounded shadow-sm" part={text}>
        <div class="d-flex gap-3">
          <div class="d-flex align-items-center">
            <img src={this.imageUrl} alt={this.modelName} height={125} width={125} part={image} />
          </div>

          <div class="d-flex flex-column flex-grow-1 gap-3 p-4">
            <div class="d-flex row flex-row">
              <div class="col-6">{this.modelName}</div>
              <div class="col-6 d-flex justify-content-end">
                <a part={link} href={this.helpUrl}>Learn more</a>
              </div>
            </div>

            <div class="row">
              <div class="col-6">{this.description}</div>
              <div class="col-6 d-flex justify-content-end align-items-center">
                <button class="p-0 minus btn" onClick={this.removeUnit} part={buttonLink}>-</button>
                <div class="w-25 text-center">{this.selectedQuantity}</div>
                <button class="p-0 plus btn" onClick={this.addUnit} part={buttonLink}>+</button>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
