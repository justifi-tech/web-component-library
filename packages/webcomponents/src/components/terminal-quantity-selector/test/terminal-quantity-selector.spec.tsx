import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { TerminalQuantitySelector } from '../terminal-quantity-selector';

describe('justifi-terminal-quantity-selector', () => {
  it('should render with all props provided', async () => {
    const page = await newSpecPage({
      components: [TerminalQuantitySelector],
      template: () => (
        <terminal-quantity-selector
          terminalId="tid_123"
          modelName="DVC100"
          imageUrl="https://device-image.com"
          helpUrl="https://learnmore.com"
          description="super duper terminal"
        />
      )
    });

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it('should emit an event when the quantity is changed and update state', async () => {
    const onSelectedQuantityChange = jest.fn();

    const page = await newSpecPage({
      components: [TerminalQuantitySelector],
      template: () => (
        <terminal-quantity-selector
          terminalId="tid_123"
          modelName="DVC100"
          imageUrl="https://device-image.com"
          helpUrl="https://learnmore.com"
          description="super duper terminal"
        />
      )
    });

    await page.waitForChanges();

    page.root.addEventListener('selectedQuantityChange', onSelectedQuantityChange);

    const addButton = page.root.querySelector('.plus');
    (addButton as HTMLElement).click();

    await page.waitForChanges();

    expect(onSelectedQuantityChange).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        terminalId: 'tid_123',
        quantity: 1
      }
    }));

    expect(page.rootInstance.selectedQuantity).toBe(1);
  });

  it('should not decrease quantity when the quantity is 0', async () => {
    const page = await newSpecPage({
      components: [TerminalQuantitySelector],
      template: () => (
        <terminal-quantity-selector
          terminalId="tid_123"
          modelName="DVC100"
          imageUrl="https://device-image.com"
          helpUrl="https://learnmore.com"
          description="super duper terminal"
        />
      )
    });

    await page.waitForChanges();

    page.rootInstance.removeUnity();

    await page.waitForChanges();

    expect(page.rootInstance.selectedQuantity).toBe(0);
  });

  it('should not add unitity when the selectedQuantity is equal to the limit', async () => {
    const page = await newSpecPage({
      components: [TerminalQuantitySelector],
      template: () => (
        <terminal-quantity-selector
          terminalId="tid_123"
          modelName="DVC100"
          imageUrl="https://device-image.com"
          helpUrl="https://learnmore.com"
          description="super duper terminal"
          limit={2}
        />
      )
    });

    await page.waitForChanges();

    const plusButton = (page.root.querySelector('.plus') as HTMLElement);
    plusButton.click();
    plusButton.click();
    plusButton.click();

    await page.waitForChanges();

    expect(page.rootInstance.selectedQuantity).toBe(2);
  });
});
