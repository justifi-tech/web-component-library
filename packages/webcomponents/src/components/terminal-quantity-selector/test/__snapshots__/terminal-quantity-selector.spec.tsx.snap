// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-terminal-quantity-selector should render with all props provided 1`] = `
<terminal-quantity-selector>
  <div class="mt-4 rounded shadow-sm" part="text color font-family">
    <div class="d-flex gap-3">
      <div>
        <img alt="DVC100" height="125" src="https://device-image.com" width="125">
      </div>
      <div class="d-flex flex-column flex-grow-1 gap-3 p-4">
        <div class="d-flex flex-row row">
          <div class="col-6">
            DVC100
          </div>
          <div class="col-6 d-flex justify-content-end">
            <a href="https://learnmore.com" part="link text color font-family">
              Learn more
            </a>
          </div>
        </div>
        <div class="row">
          <div class="col-6">
            super duper terminal
          </div>
          <div class="col-6 d-flex justify-content-end">
            <button class="minus p-0" part="button-link button text color font-family">
              -
            </button>
            <div class="text-center w-25">
              0
            </div>
            <button class="p-0 plus" part="button-link button text color font-family">
              +
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</terminal-quantity-selector>
`;
